### ISSUE TITLE
RFC: `No goals` should point to errors elsewhere

### ISSUE THREAD

**Issue by fpvandoorn** [leanprover/lean4#4190]
### Proposal

The current `No goals` message is confusing for beginners. They see a `No goals` message, and miss that there are errors in other places, and think they have finished a proof, even though there are still errors. I have seen this with multiple beginners, and have even missed an error myself.

Example:
```lean
example : True ∧ True ∧ True := by
  refine ⟨?_, ?_, ?_⟩
  · have h : True := trivial
  · trivial
    
--  ^ -- if you put your cursor here, you get a "No goals" message, and only three characters in the entire file have squiggly underlines, which are easy to miss.
```

Note: there are various workarounds you can look at. The `All messages` tab in the Infoview, using the VSCode `ErrorLens` extension, ending your proof with the `done` tactic (which only allows you to see the outermost goal, not intermediates, unless you use `done` there, too).

**Proposal** In my ideal world: 
* the `No goals` message also displays the location of all errors in the type or value of your current declaration (mostly importantly the unsolved goal errors, but also other errors).
* These locations are clickable which will move your cursor to them
* When jumping to a `unsolved goal` error, it should go to the *end* of the sequence of tactics for that goal, adding a newline + correct indentation if there is no newline in that position yet
* If there are no other errors in that declaration, it will show a nicer message that you have finished the proof / declaration (e.g. **Goals Accomplished 🎉**).

  - **User Experience**: Improved by making it easier to see where the errors are in a declaration.

  - **Beneficiaries**: Beginners / everyone

  - **Maintainability**: Unsure. It might require extra data in places (the tactic state?).

### Related RFCs

* #1970 also asks to distinguish between `No goals` and `Goals accomplished`.
* #4181 asks to not show `No Goals` with certain errors.

### Community Feedback

[Zulip](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/.60no.20goals.60.20should.20point.20the.20user.20to.20errors.20elsewhere)

### Impact

Add :+1: to [issues you consider important](https://github.com/leanprover/lean4/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc). If others benefit from the changes in this proposal being added, please ask them to add :+1: to it.

**Comment by nomeata** [comment]
I have a vision (which keeps getting preempted) that an incomplete proof isn't shown by squiggly lines at the `by` or the `\.` at the beginning of the incomplete tactic sequence, but by a symbol like 🚧 or ⚠️ at the end, i.e. at the position where the user ought to continue typing. This may alleviate some of the issues you describe. Looks like I never wrote it down in an issue though. It's a bit blocked on https://github.com/leanprover/lean4/pull/3910 because it likely needs to use LSP inlay hints.

**Comment by fpvandoorn** [comment]
Oh, I think that would be a great improvement, and might indeed solve all the problems this RFC is trying to solve.

**Comment by fpvandoorn** [comment]
One issue that is not be solved by your suggestion is when you have a (non-unsolved goal) error in your declarations. That can still be easy to miss. And certain errors prevent other errors / messages from showing (cf. #3556). But maybe you can annotate other errors with "❌" in the source file?

**Comment by mhuisi** [comment]
@fpvandoorn Do you think that these issues have been sufficiently resolved by the new editor decorations?

**Comment by fpvandoorn** [comment]
Yes, I think most of this issue has been resolved. I haven't played with Lean much yet this month, but I have one remark:

Would it be possible to put the "work in progress" icon in the exact place where the (next) tactic for that goal is expected?

![Image](https://github.com/user-attachments/assets/269a31d7-e874-4463-998c-dfc149fdf730)

For example, in the image above; if I type where the icon is, I'll get an error that the first goal has already been solved.

I guess this request is hard if the next line is non-empty...

**Comment by mhuisi** [comment]
I played around with that idea while designing the feature, but it is very difficult to get right and there are a number of technical challenges. There are also some inherent problems with this idea that I have no great solution for. The next line being non-empty is one of them, but another is that once the semantics of the symbol become "this is where to continue typing" instead of the much simpler "an 'unsolved goals' error ends on this line", we also need to be able to display the symbol in the middle of a line, e.g. as in the following example:
```
example : True ∧ True := And.intro (by ) sorry
```

Now, as with inlay hints, the symbol can push the text following it to the left or to the right when it pops up, disappears or moves around, which can be very annoying. For inlay hints, we work around this with a 3000ms edit delay, as they do not need to be updated that quickly. But prior testing has shown that 3000ms of edit delay would be far too much for the 'unsolved goals' symbols, as you usually want to see them relatively quickly, and that something like 750ms is sufficient since the symbol can only move around at the end of lines.

**Comment by fpvandoorn** [comment]
Makes sense. I think the original issue here is (mostly) solved.
